### ISSUE TITLE
EntitySqlException when calling LoadRelatedEntitiesAsync

### ISSUE THREAD

**Issue by weitzhandler** [TrackableEntities/trackable-entities#165]
Hi,

Here's the relevant part of my model:

```
public partial class Runsheet : ITrackable
{                                                          
  public int Id { get; set; }
  public virtual ICollection<Patient> Patients { get; set; }
} 
public partial class Patient : ITrackable
{
  public int Id { get; set; }
  [ForeignKey(nameof(Pcr))]
  public int? PcrId { get; set; }
  [InverseProperty(nameof(Models.Pcr.Patient))]
  public virtual Pcr Pcr { get; set; }
}
public partial class Pcr : ITrackable
{
  [Key, ForeignKey(nameof(Patient))]
  public int PatientId { get; set; }  
  public virtual Patient Patient { get; set; }
}
```

Here's what my controller looks like:

    [HttpPut]
    public async Task<IActionResult> Put([FromBody] Runsheet runsheet)
    {
      DbContext.ApplyChanges(runsheet);
      try
      {
        await DbContext.SaveChangesAsync();
      }
      catch (DbUpdateConcurrencyException ex)
      {
        if (!DbContext.Runsheets.Any(r => r.Id == runsheet.Id))
          return StatusCode((int)HttpStatusCode.Conflict, ex);
      }

      //This line causes the problem
      await DbContext.LoadRelatedEntitiesAsync(runsheet);
      runsheet.AcceptChanges();
      return Ok(runsheet);
    }

><code>System.Data.Entity.Core.EntitySqlException: 'Id' is not a member of type 'MyProj.Models.Pcr' in the currently loaded schemas. Near simple identifier, line 1, column 52.
   at System.Data.Entity.Core.Common.EntitySql.SemanticResolver.ResolvePropertyAccess(DbExpression valueExpr, String name, ErrorContext errCtx)
   at System.Data.Entity.Core.Common.EntitySql.SemanticAnalyzer.ConvertDotExpr(Node expr, SemanticResolver sr)
   at System.Data.Entity.Core.Common.EntitySql.SemanticAnalyzer.Convert(Node astExpr, SemanticResolver sr)
   at System.Data.Entity.Core.Common.EntitySql.SemanticAnalyzer.ConvertValueExpressionAllowUntypedNulls(Node astExpr, SemanticResolver sr)
   at System.Data.Entity.Core.Common.EntitySql.SemanticAnalyzer.ConvertInExprArgs(BuiltInExpr astBuiltInExpr, SemanticResolver sr)
   at System.Data.Entity.Core.Common.EntitySql.SemanticAnalyzer.<CreateBuiltInExprConverter>b__92(BuiltInExpr bltInExpr, SemanticResolver sr)
   at System.Data.Entity.Core.Common.EntitySql.SemanticAnalyzer.ConvertBuiltIn(Node astExpr, SemanticResolver sr)
   at System.Data.Entity.Core.Common.EntitySql.SemanticAnalyzer.Convert(Node astExpr, SemanticResolver sr)
   at System.Data.Entity.Core.Common.EntitySql.SemanticAnalyzer.ConvertValueExpressionAllowUntypedNulls(Node astExpr, SemanticResolver sr)
   at System.Data.Entity.Core.Common.EntitySql.SemanticAnalyzer.ConvertValueExpression(Node astExpr, SemanticResolver sr)
   at System.Data.Entity.Core.Common.EntitySql.SemanticAnalyzer.ProcessWhereHavingClausePredicate(DbExpressionBinding source, Node predicate, ErrorContext errCtx, String bindingNameTemplate, SemanticResolver sr)
   at System.Data.Entity.Core.Common.EntitySql.SemanticAnalyzer.ConvertQueryExpr(Node expr, SemanticResolver sr)
   at System.Data.Entity.Core.Common.EntitySql.SemanticAnalyzer.Convert(Node astExpr, SemanticResolver sr)
   at System.Data.Entity.Core.Common.EntitySql.SemanticAnalyzer.ConvertValueExpressionAllowUntypedNulls(Node astExpr, SemanticResolver sr)
   at System.Data.Entity.Core.Common.EntitySql.SemanticAnalyzer.ConvertQueryStatementToDbExpression(Statement astStatement, SemanticResolver sr, List`1& functionDefs)
   at System.Data.Entity.Core.Common.EntitySql.SemanticAnalyzer.AnalyzeQueryCommand(Node astExpr)
   at System.Data.Entity.Core.Common.EntitySql.CqlQuery.<AnalyzeQueryExpressionSemantics>b__8(SemanticAnalyzer analyzer, Node astExpr)
   at System.Data.Entity.Core.Common.EntitySql.CqlQuery.AnalyzeSemanticsCommon[TResult](Node astExpr, Perspective perspective, ParserOptions parserOptions, IEnumerable`1 parameters, IEnumerable`1 variables, Func`3 analysisFunction)
   at System.Data.Entity.Core.Common.EntitySql.CqlQuery.AnalyzeQueryExpressionSemantics(Node astQueryCommand, Perspective perspective, ParserOptions parserOptions, IEnumerable`1 parameters, IEnumerable`1 variables)
   at System.Data.Entity.Core.Common.EntitySql.CqlQuery.<>c__DisplayClass4.<CompileQueryCommandLambda>b__3(Node astCommand, ParserOptions validatedParserOptions)
   at System.Data.Entity.Core.Common.EntitySql.CqlQuery.CompileCommon[TResult](String commandText, ParserOptions parserOptions, Func`3 compilationFunction)
   at System.Data.Entity.Core.Common.EntitySql.CqlQuery.CompileQueryCommandLambda(String queryCommandText, Perspective perspective, ParserOptions parserOptions, IEnumerable`1 parameters, IEnumerable`1 variables)
   at System.Data.Entity.Core.Objects.EntitySqlQueryState.Parse()
   at System.Data.Entity.Core.Objects.EntitySqlQueryState.GetExecutionPlan(Nullable`1 forMergeOption)
   at System.Data.Entity.Core.Objects.ObjectQuery`1.<>c__DisplayClassc.<GetResultsAsync>b__a()
   at System.Data.Entity.Core.Objects.ObjectContext.<ExecuteInTransactionAsync>d__3d`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Data.Entity.SqlServer.DefaultSqlExecutionStrategy.<ExecuteAsyncImplementation>d__9`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Data.Entity.Utilities.TaskExtensions.CultureAwaiter`1.GetResult()
   at System.Data.Entity.Core.Objects.ObjectQuery`1.<GetResultsAsync>d__e.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at TrackableEntities.EF6.DbContextExtensions.<ExecuteQueryEntitySqlAsync>d__42.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at TrackableEntities.EF6.DbContextExtensions.<GetRelatedEntitiesAsync>d__27.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at TrackableEntities.EF6.DbContextExtensions.<LoadRelatedEntitiesAsync>d__13.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at TrackableEntities.EF6.DbContextExtensions.<LoadRelatedEntitiesOnPropertiesAsync>d__14.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at TrackableEntities.EF6.DbContextExtensions.<LoadRelatedEntitiesAsync>d__13.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at MyProj.Controllers.RunsheetsController.<Put>d__4.MoveNext()
</code>

**Comment by tonysneed** [comment]
What is the relationship between Patient and Pcr?

**Comment by weitzhandler** [comment]
I copied the model to my post.

It's a Patient(1)<->Pcr(0..1)

**Comment by tonysneed** [comment]
It looks like there is something wrong with your model of a **one-to-zero-or-one** relationship between Patient and Pcr.  Patient could not have a PcrId property.  See this [article](http://www.entityframeworktutorial.net/code-first/configure-one-to-one-relationship-in-code-first.aspx) for a good example of how to model this kind of relationship.

Once you get the model right, we can see if `LoadRelatedEntitiesAsync` is populating the reference properties of the related entities.

**Comment by weitzhandler** [comment]
Thanks @tonysneed 
I changed my model to the following:

  public partial class Patient : IValidatableObject
  {
    public int Id { get; set; }
    public virtual Pcr Pcr { get; set; }
  }

  public partial class Pcr
  {
    [Key, ForeignKey(nameof(Patient))]
    public int PatientId { get; set; }  
    public virtual Patient Patient { get; set; }
  }

And I'm still getting the same exception:

>'Id' is not a member of type 'Hatzoloh.Models.Pcr' in the currently loaded schemas. Near simple identifier, line 1, column 52.

[Full StackTrace](https://hastebin.com/cefegumesu.css)

**Comment by tonysneed** [comment]
As an experiment, can you rename PatientId to Id in Pcr?

**Comment by weitzhandler** [comment]
Well that eliminates the issue. I'd still want to stick to my original naming tho.

**Comment by tonysneed** [comment]
Keep in mind that LRE is there purely for convenience and that you can instead use the EF LoadProperty method to manually populate reference properties.

**Comment by weitzhandler** [comment]
Do you know what makes it fail if the identity column isn't named `Id`? Can we rely on the `Key` attribute in addition to naming convention?

**Comment by tonysneed** [comment]
Have a look at `GetRelatedEntitiesSql` [here](https://github.com/TrackableEntities/trackable-entities/blob/develop/Source/TrackableEntities.EF.5/DbContextExtensions.cs#L771).  I would suggest creating a failing test, which you can then debug to see what's going on.

**Comment by tonysneed** [comment]
You can also submit a PR with a failing test, which will make it easier for me to debug when I have time.

**Comment by weitzhandler** [comment]
Apparently this has to do with non-mapped (#192) and `[JsonIgnore]`d properties (#184).

**Comment by weitzhandler** [comment]
@tonysneed, I'm still getting an exception when calling `LoadRelatedEntitiesAsync`.

I get an `EntitySqlException`:

> The query syntax is not valid. Near term ',', line 1, column 68.

Stack trace:

>    at System.Data.Entity.Core.Common.EntitySql.CqlParser.yyerror(String s)
>    at System.Data.Entity.Core.Common.EntitySql.CqlParser.yyparse()
>    at System.Data.Entity.Core.Common.EntitySql.CqlParser.internalParseEntryPoint()
>    at System.Data.Entity.Core.Common.EntitySql.CqlParser.Parse(String query)
>    at System.Data.Entity.Core.Common.EntitySql.CqlQuery.Parse(String commandText, ParserOptions parserOptions)
>    at System.Data.Entity.Core.Common.EntitySql.CqlQuery.CompileCommon[TResult](String commandText, ParserOptions parserOptions, Func`3 compilationFunction)
>    at System.Data.Entity.Core.Common.EntitySql.CqlQuery.CompileQueryCommandLambda(String queryCommandText, Perspective perspective, ParserOptions parserOptions, IEnumerable`1 parameters, IEnumerable`1 variables)
>    at System.Data.Entity.Core.Objects.EntitySqlQueryState.Parse()
>    at System.Data.Entity.Core.Objects.EntitySqlQueryState.GetExecutionPlan(Nullable`1 forMergeOption)
>    at System.Data.Entity.Core.Objects.ObjectQuery`1.<>c__DisplayClassc.<GetResultsAsync>b__a()
>    at System.Data.Entity.Core.Objects.ObjectContext.<ExecuteInTransactionAsync>d__3d`1.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
>    at System.Data.Entity.SqlServer.DefaultSqlExecutionStrategy.<ExecuteAsyncImplementation>d__9`1.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
>    at System.Data.Entity.Utilities.TaskExtensions.CultureAwaiter`1.GetResult()
>    at System.Data.Entity.Core.Objects.ObjectQuery`1.<GetResultsAsync>d__e.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
>    at TrackableEntities.EF6.DbContextExtensions.<ExecuteQueryEntitySqlAsync>d__7d.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
>    at TrackableEntities.EF6.DbContextExtensions.<GetRelatedEntitiesAsync>d__57.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
>    at TrackableEntities.EF6.DbContextExtensions.<LoadRelatedEntitiesAsync>d__2f.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
>    at TrackableEntities.EF6.DbContextExtensions.<LoadRelatedEntitiesOnPropertiesAsync>d__3e.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
>    at TrackableEntities.EF6.DbContextExtensions.<LoadRelatedEntitiesAsync>d__2f.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
>    at MyProject.Controllers.MyDataTypesController.<Put>d__6.MoveNext() in D:\Users\Shimmy\Documents\Visual Studio 2017\Projects\MyProject\MyProject.Api\Controllers\MyDataTypesController.cs:line 278

**Comment by tonysneed** [comment]
Yes, I can make it more robust with something like an onError callback.

**Comment by weitzhandler** [comment]
Would be nice if we can choose to continue loading the rest of the related entities even an error is caught, maybe make the `onError` delegate provide a value for whether to continue or abort etc.

**Comment by tonysneed** [comment]
Yes I’ll make it a Func.

**Comment by weitzhandler** [comment]
Consider `Action<LoadRelatedEntitiesError>`, where the it has a property that is read when back in the LRE function that indicates what to do next.

**Comment by tonysneed** [comment]
@weitzhandler Can you give me some info on the model you're using when you get the exception on LRE?  I need to generate an error to test.

**Comment by tonysneed** [comment]
This is now resolved by #215.  For usage, see [unit tests](https://github.com/TrackableEntities/trackable-entities/blob/dd27a7220a94626c6c05300c359ec8fa0cb6cf72/Source/Tests/TrackableEntities.EF.5.Tests/LoadRelatedEntitiesTests.cs#L961):
- LoadRelatedEntitiesAsync_OnLoading_Should_Continue
- LoadRelatedEntitiesAsync_OnLoading_Should_Replace
- LoadRelatedEntitiesAsync_OnError_Should_Throw
- LoadRelatedEntitiesAsync_OnError_Should_Continue

The `onLoading` parameter allows you to skip or replace the built-in entity loading by loading the entities manually.  The `onError` parameter also allows you to load entities manually, then return true to continue execution without throwing an exception.

**Comment by weitzhandler** [comment]
Thanks to your consistent dedication and your detailed explanations here and elsewhere, I managed to fix all the issues I had and it's all working like a charm now. You're GREAT!

Thank you from the bottom of my heart @tonysneed!
